### PR TITLE
libultra voice OK

### PR DESCRIPTION
**Pull request by engineer124**
A special thanks to @EllipticEllipsis @hensldm @Thar0 for the help on these functions and documentation. For those who work on libultra, it would be great to get feedback for styling/docs that align with more broad libultra efforts!

### CODE REVIEW

File: src/code/z_message.c [hunk 1]
```diff
@@ -240,7 +241,7 @@ void func_8014CCB4(PlayState* play, s16* decodedBufPos, s32* offset, f32* arg3)
     k += FONT_CHAR_TEX_SIZE;
     msgCtx->decodedBuffer.wchar[t] = 0x8373;
     t += 1;
-    Font_LoadChar(play, 0x815C, k); // Ox815C = ― in JISX0213
+    Font_LoadChar(play, 0x815C, k); // 0x815C = ― in JISX0213
```
[the remark this answers is not in the document]
Comment (engineer124) [reply src/code/z_message.c:244]: The letter "O" -> the number "0"

[the remark this answers is not in the document]
Comment (AngheloAlf) [reply src/code/z_message.c:244]: O
i see, thanks

File: include/ultra64/controller_voice.h [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+s32 __osVoiceContRead2(OSMesgQueue* mq, s32 channel, u16 address, u8 dst[2]);
+s32 __osVoiceContRead36(OSMesgQueue* mq, s32 channel, u16 address, u8 dst[36]);
+s32 __osVoiceContWrite4(OSMesgQueue* mq, s32 channel, u16 address, u8 dst[4]);
+s32 __osVoiceContWrite20(OSMesgQueue* mq, s32 channel, u16 address, u8 dst[20]);
+s32 __osVoiceCheckResult(OSVoiceHandle* hd, u8* status);
+s32 __osVoiceGetStatus(OSMesgQueue* mq, s32 channel, u8* status);
+s32 __osVoiceSetADConverter(OSMesgQueue* mq, s32 channel, u8 data);
+u8 __osVoiceContDataCrc(u8* data, size_t numBytes);
+
+extern OSVoiceHandle gVoiceHandle;
```
[the remark this answers is not in the document]
Comment (engineer124) [reply include/ultra64/controller_voice.h]: Oh yeah I suppose, I just associated the type with `OSVoiceHandle`. I put the extern in `z64audio.h` for now

File: tools/disasm/variables.txt [hunk 3]
```diff
@@ -71,9 +71,7 @@
     0x80097F70:("__osTimerList","OSTimer*","",0x4),
     0x80097F80:("D_80097F80","f32","",0x4),
     0x80097F90:("D_80097F90","f32","",0x4),
-    0x80097FA0:("D_80097FA0","UNK_TYPE1","",0x1),
-    0x80097FA4:("D_80097FA4","UNK_TYPE1","",0x1),
-    0x80097FA5:("D_80097FA5","UNK_TYPE1","",0x1),
+    0x80097FA0:("sCmd","UNK_TYPE1","",0x1),
```
Comment (hensldm) [inline tools/disasm/variables.txt:74]:
```suggestion
    0x80097FA0:("sCmds","UNK_TYPE1","",0x1),
```

File: src/libultra/voice/voicesetadconverter.c [hunk 4]
```diff
@@ -1,3 +1,70 @@
...
+ * File: voicesetadconverter.c
+ */
 
-#pragma GLOBAL_ASM("asm/non_matchings/boot/voicesetadconverter/__osVoiceSetADConverter.s")
+#include "ultra64/controller_voice.h"
+#include "io/controller.h"
+#include "functions.h"
+#include "variables.h"
+
+#define SWRITEFORMAT(p) ((__OSVoiceSWriteFormat*)(p))
```
Comment (hensldm) [inline src/libultra/voice/voicesetadconverter.c:10]: Oh I must have missed this (looking at it it is all the formats I added). The arg should probably be `ptr` not just `p`. That way it mirrors the `READFORMAT` in `controller.h`.

File: src/libultra/voice/voicecontread2.c [hunk 5]
```diff
@@ -1,3 +1,74 @@
...
+ * File: voicecontread2.c
+ */
 
-#pragma GLOBAL_ASM("asm/non_matchings/boot/voicecontread2/__osVoiceContRead2.s")
+#include "ultra64/controller_voice.h"
+#include "io/controller.h"
+#include "functions.h"
+#include "variables.h"
+
+#define READ2FORMAT(p) ((__OSVoiceRead2Format*)(ptr))
```
Comment (hensldm) [inline src/libultra/voice/voicecontread2.c:10]: This one as well.

File: src/libultra/voice/voicecontread36.c [hunk 6]
```diff
@@ -1,3 +1,74 @@
...
+ * File: voicecontread36.c
+ */
 
-#pragma GLOBAL_ASM("asm/non_matchings/boot/voicecontread36/__osVoiceContRead36.s")
+#include "ultra64/controller_voice.h"
+#include "io/controller.h"
+#include "functions.h"
+#include "variables.h"
+
+#define READ36FORMAT(p) ((__OSVoiceRead36Format*)(ptr))
```
Comment (hensldm) [inline src/libultra/voice/voicecontread36.c:10]: Same here

File: src/libultra/voice/voicecontwrite20.c [hunk 7]
```diff
@@ -1,3 +1,75 @@
...
+ * File: voicecontwrite20.c
+ */
 
-#pragma GLOBAL_ASM("asm/non_matchings/boot/voicecontwrite20/__osVoiceContWrite20.s")
+#include "ultra64/controller_voice.h"
+#include "io/controller.h"
+#include "functions.h"
+#include "variables.h"
+
+#define WRITE20FORMAT(p) ((__OSVoiceWrite20Format*)(ptr))
```
Comment (hensldm) [inline src/libultra/voice/voicecontwrite20.c:10]: Same here

File: src/libultra/voice/voicecontwrite4.c [hunk 8]
```diff
@@ -1,3 +1,74 @@
...
+ * File: voicecontwrite4.c
+ */
 
-#pragma GLOBAL_ASM("asm/non_matchings/boot/voicecontwrite4/__osVoiceContWrite4.s")
+#include "ultra64/controller_voice.h"
+#include "io/controller.h"
+#include "functions.h"
+#include "variables.h"
+
+#define WRITE4FORMAT(p) ((__OSVoiceWrite4Format*)(ptr))
```
Comment (hensldm) [inline src/libultra/voice/voicecontwrite4.c:10]: Same here.

File: src/libultra/voice/voicecontread2.c (hunk 5 above)
Comment (engineer124) [reply src/libultra/voice/voicecontread2.c]: How does this still work when `p` is different than `ptr`?

Comment (hensldm) [reply src/libultra/voice/voicecontread2.c]: Confused me as well, than I realized the local variable was called `ptr`. So the macro arg just gets dropped, but the `ptr` matches the local variable so it still works.

**Review by hensldm — APPROVED** [review]

File: src/libultra/voice/voicecontrolgain.c [hunk 9]
```diff
@@ -1,5 +1,41 @@
...
+u8 sDigitalTable[] = { 0x00, 0x80, 0x40, 0xC0, 0x20, 0xA0, 0x60, 0xE0 };
+
+s32 osVoiceControlGain(OSVoiceHandle* hd, s32 analog, s32 digital) {
+    s32 errorCode;
+    u8 cmd;
+
+    if (analog == 0) {
+        cmd = 0x18;
+    } else {
+        cmd = 0x98;
```
Comment (EllipticEllipsis) [inline src/libultra/voice/voicecontrolgain.c:21]: Do these have defines in controller.h? If not they probably should.

Comment (engineer124) [reply src/libultra/voice/voicecontrolgain.c:21]: https://ultra64.ca/files/documentation/online-manuals/man/pro-man/pro26/26-08.html#osVoiceControlGain()
image

Any suggestions? Would these command names be known or would we have to pick a reasonable name ourselves?

Comment (engineer124) [reply src/libultra/voice/voicecontrolgain.c:21]: Similarly for the digital: 
`cmd = sDigitalTable[digital] + 2;`
`u8 sDigitalTable[] = { 0x00, 0x80, 0x40, 0xC0, 0x20, 0xA0, 0x60, 0xE0 };`
(I wonder about the +2)
image

File: src/libultra/voice/voicesetadconverter.c [hunk 10]
```diff
@@ -1,3 +1,70 @@
...
+            __osPfsLastChannel = channel;
+
+            for (i = 0; i < channel; i++, *ptr++ = 0) {}
+
+            __osPfsPifRam.status = CONT_CMD_EXE;
+
+            ptr[0] = CONT_CMD_SWRITE_VOICE_TX;
+            ptr[1] = CONT_CMD_SWRITE_VOICE_RX;
+            ptr[2] = CONT_CMD_SWRITE_VOICE;
+            ptr[5] = 0;
```
Comment (hensldm) [inline src/libultra/voice/voicesetadconverter.c:35]: Oh I missed these. They should probably use the SWRITEFORMAT macro and set the appropriate struct fields.

**Review by jpburnett — APPROVED** [review]
🎤

File: src/libultra/voice/voicecontrolgain.c (hunk 9 above)
Comment (engineer124) [reply src/libultra/voice/voicecontrolgain.c:21]: Bump on this. Any libultra folks have a name suggestions? Can always leave it to later as well

Comment (hensldm) [reply src/libultra/voice/voicecontrolgain.c:21]: Fwiw, ultralib has just left them as hex for the time being, though defines sound fine for now (though that +2 is still weird).
Would probably also want to do the same for the `sCmds` array in `voiceinit`.

**Pull request merged**